### PR TITLE
Allow custom identity driver

### PR DESCRIPTION
**Pull request by mnederlof**
(which falls back to sql) and allow custom configuration by defining keystone.server.custom_sections

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -330,6 +330,21 @@ Enable Federated keystone
...
+
+.. code-block:: yaml
+
+    keystone:
+      server:
+        identity:
+          driver = custom_driver
+        custom_sections:
+          custom_driver:
+            my_option = my_value
```
Comment (fpytloun) [inline README.rst]: This is not valid yaml afaik

File: keystone/files/mitaka/keystone.conf.Debian [hunk 2]
```diff
@@ -2193,3 +2193,11 @@ driver = memcache_pool
 
 [extra_headers]
 Distribution = Ubuntu
+
+{% for section in server.get('custom_sections', {}) %}
+[{{ section }}]
```
Comment (fpytloun) [inline keystone/files/mitaka/keystone.conf.Debian]: This is definately not what we want to do - allow user to define anything without any validation etc.
Create specific metadata for what you are trying to achieve.

**Review by fpytloun — CHANGES_REQUESTED** [review]

File: README.rst (hunk 1 above)
Comment (mnederlof) [reply README.rst]: good catch; now updated.

File: keystone/files/mitaka/keystone.conf.Debian (hunk 2 above)
Comment (mnederlof) [reply keystone/files/mitaka/keystone.conf.Debian]: I'll update it with new data.

**Comment by mnederlof** [comment]
@fpytloun : i have updated the configuration to reflect what i need to achieve. Is this what you meant?

File: keystone/files/mitaka/keystone.conf.Debian [hunk 3]
```diff
@@ -951,7 +951,7 @@ domain_config_dir = /etc/keystone/domains
 {%- if server.get('backend', 'sql') == 'ldap' %}
 driver = ldap
 {%- else %}
-driver = sql
+driver = {{ server.get('identity', {}).get('backend', 'sql') }}
```
Comment (fpytloun) [inline keystone/files/mitaka/keystone.conf.Debian]: It should be `server.get('backend', 'sql')` as above

File: keystone/files/mitaka/keystone.conf.Debian [hunk 4]
```diff
@@ -2193,3 +2193,52 @@ driver = memcache_pool
 
 [extra_headers]
 Distribution = Ubuntu
+
+{% if server.get('k2k') %}
```
Comment (fpytloun) [inline keystone/files/mitaka/keystone.conf.Debian]: Check ldap part with include _ldap.conf, you want to do something very similar.
This won't work if k2k is not defined etc.:
```{%- if server.k2k is defined %}```

File: keystone/files/mitaka/keystone.conf.Debian [hunk 5]
```diff
@@ -2193,3 +2193,52 @@ driver = memcache_pool
...
+
+#
+# Specific driver configuration for identity backend k2k
+# (Authentication against other keystone backend as fallback)
+#
+
+# Authentication URL of keystone to authenticate against (v2.0 only)
+# (string value)
+#auth_url = https://keystone.example.com/v2.0
+auth_url = {{ server.k2k.get('auth_url')}}
```
Comment (fpytloun) [inline keystone/files/mitaka/keystone.conf.Debian]: No need to call .get unless you want to define default.

File: keystone/files/mitaka/keystone.conf.Debian (hunk 3 above)
Comment (mnederlof) [reply keystone/files/mitaka/keystone.conf.Debian]: Maybe remove the if/else around it as well then?

File: keystone/files/mitaka/keystone.conf.Debian [hunk 6]
```diff
@@ -2193,3 +2193,52 @@ driver = memcache_pool
...
+id_prefix = {{ server.k2k.get('id_prefix', 'k2k:')}}
+
+# This value will be set in the User object after authentication (string value)
+#domain = default
+domain = {{ server.k2k.get('domain', 'default')}}
+
+# Toggle for k2k caching. This has no effect unless global caching is enabled.
+# (boolean value)
+#caching = true
+caching = {{ server.k2k.get('caching', 'true')}}
```
Comment (fpytloun) [inline keystone/files/mitaka/keystone.conf.Debian]: You should use this to allow booleans:
```caching = {{ server.k2k.get('caching', True)}|lower }}```

File: keystone/files/mitaka/keystone.conf.Debian (hunk 3 above)
Comment (fpytloun) [reply keystone/files/mitaka/keystone.conf.Debian]: yes

Comment (mnederlof) [reply keystone/files/mitaka/keystone.conf.Debian]: Or only specify the options which are supported by the salt-formula.. (like an if/elseif/else structure there)

File: keystone/files/mitaka/keystone.conf.Debian (hunk 4 above)
Comment (mnederlof) [reply keystone/files/mitaka/keystone.conf.Debian]: will adapt :)

**Comment by mnederlof** [comment]
@fpytloun thanks for your feedback 👍 I have updated the configurations again. I believe this is more in line with the rest of the formula.

File: README.rst [hunk 7]
```diff
@@ -330,6 +330,26 @@ Enable Federated keystone
           - wsgi
           - shib2
 
+Use a custom identity driver with custom options
+
+.. code-block:: yaml
+
+    keystone:
+      server:
+        identity:
```
Comment (fpytloun) [inline README.rst]: Cool, almost there :-)
Just fix this part of readme, it's only this:
```
server:
  backend: k2k
```

**Comment by mnederlof** [comment]
Updated :)

File: README.rst [hunk 8]
```diff
@@ -330,6 +330,25 @@ Enable Federated keystone
           - wsgi
           - shib2
 
+Use a custom identity driver with custom options
+
+.. code-block:: yaml
+
+    keystone:
+      server:
+        driver: k2k
```
Comment (fpytloun) [inline README.rst]: backend, not driver? :-)

Comment (mnederlof) [reply README.rst]: ehm, yes :-D updated in a new commit.

**Pull request merged**